packages/core/src/validate.ts

```diff
@@ -134,15 +166,51 @@ export function getStorageLayout(validation: Validation, version: Version): Stor
   return layout;
 }
 
-export function assertUpgradeSafe(validation: Validation, version: Version, unsafeAllowCustomTypes = false): void {
+function getUnlinkedBytecode(validation: Validation, bytecode: string): string {
+  let unlinkedBytecode: string = bytecode;
+  Object.keys(validation)
+    .filter(name => validation[name].linkReferences.length > 0)
+    .find(name => {
+      const linkReferences: LinkReference[] = getLinkReferences(validation, name);
+      unlinkedBytecode = replaceLinkReferences(bytecode, linkReferences);
+      const version = getVersion(unlinkedBytecode);
+      validation[name].version?.withMetadata === version.withMetadata;
+    });
+  return unlinkedBytecode;
+}
+
+function getLinkReferences(validation: Validation, contractName: string): LinkReference[] {
+  const c = validation[contractName];
+
+  return c.linkReferences
+    .concat(...c.inherit.map(name => validation[name].linkReferences))
+    .concat(...c.libraries.map(name => validation[name].linkReferences));
+}
+
+export function assertUpgradeSafe(validation: Validation, version: Version, opts: ValidationOptions): void {
   const contractName = getContractName(validation, version);
   let errors = getErrors(validation, version);
   let thereAreExceptions = false;
+  let exceptionMessage = '';
```

Comment:
Instead of concatenating all warning messages into a single warning, I would like this to print the multiple warnings separately, each with its own "Warning" label. This is to make it stand out as much as possible.

Reply:
Done.